src/server/pps/server/pipeline_controller.go

```diff
@@ -242,95 +356,98 @@ func (pc *pipelineController) fetchState() (*pps.PipelineInfo, context.Context,
+			// default: scale up if pipeline start hasn't propagated to the collection yet
+			// Note: mostly this should do nothing, as this runs several times per job
+			ScaleWorkersSideEffect(sideEffectToggle_UP),
+		}, nil
 	case pps.PipelineState_PIPELINE_STANDBY:
-		pc.stopCrashingPipelineMonitor()
-		// Make sure pipelineMonitor is running to pull it out of standby
-		pc.startPipelineMonitor(pi)
-		// default: scale down if standby hasn't propagated to kube RC yet
-		return pc.scaleDownPipeline(ctx, pi, rc)
+		return pi.State, []sideEffect{
+			CrashingMonitorSideEffect(sideEffectToggle_DOWN),
+			// Make sure pipelineMonitor is running to pull it out of standby
+			PipelineMonitorSideEffect(sideEffectToggle_UP),
+			// default: scale down if standby hasn't propagated to kube RC yet
+			ScaleWorkersSideEffect(sideEffectToggle_DOWN),
+		}, nil
 	case pps.PipelineState_PIPELINE_CRASHING:
-		// start a monitor to poll k8s and update us when it goes into a running state
-		pc.startPipelineMonitor(pi)
-		pc.startCrashingPipelineMonitor(pi)
-		// Surprisingly, scaleUpPipeline() is necessary, in case a pipelines is
-		// quickly transitioned to CRASHING after coming out of STANDBY. Because the
-		// pipeline controller reads the current state of the pipeline after each
-		// event (to avoid getting backlogged), it might never actually see the
-		// pipeline in RUNNING. However, if the RC is never scaled up, the pipeline
-		// can never come out of CRASHING, so do it here in case it never happened.
-		//
-		// In general, CRASHING is actually almost identical to RUNNING (except for
-		// the monitorCrashing goro)
-		return pc.scaleUpPipeline(ctx, pi, rc)
+		return pi.State, []sideEffect{
+			// start a monitor to poll k8s and update us when it goes into a running state
+			CrashingMonitorSideEffect(sideEffectToggle_UP),
+			PipelineMonitorSideEffect(sideEffectToggle_UP),
+			// Surprisingly, scaleUpPipeline() is necessary, in case a pipelines is
+			// quickly transitioned to CRASHING after coming out of STANDBY. Because the
+			// pipeline controller reads the current state of the pipeline after each
+			// event (to avoid getting backlogged), it might never actually see the
+			// pipeline in RUNNING. However, if the RC is never scaled up, the pipeline
+			// can never come out of CRASHING, so do it here in case it never happened.
+			//
+			// In general, CRASHING is actually almost identical to RUNNING (except for
+			// the monitorCrashing goro)
+			ScaleWorkersSideEffect(sideEffectToggle_UP),
+		}, nil
+	}
+	return 0, nil, errors.New("could not evaluate pipeline transition")
+}
+
+func (pc *pipelineController) apply(ctx context.Context, pi *pps.PipelineInfo, rc *v1.ReplicationController, target pps.PipelineState, sideEffects []sideEffect) error {
+	for _, s := range sideEffects {
+		if err := s.apply(ctx, pc, pi, rc); err != nil {
+			return err
+		}
+	}
+	if target != pi.State {
+		return pc.setPipelineState(ctx, pi.SpecCommit, target, "")
```

Comment:
I'm torn between returning the "desired" pipeline state, as you're doing here, and making this explicitly about a change, where you check against some explicit "don't care" or "same" value which gets returned instead of `pi.State` in all of those places. I think the desired state is better, but it would be good to eliminate the weirdness about the restarts in that case.

Part of the issue is that `setPipelineState` itself has the side effect of causing a bump. Avoiding the same start/end case will be less important if/when we remove the cases where we rely on multiple "steps" to completely resolve an update.